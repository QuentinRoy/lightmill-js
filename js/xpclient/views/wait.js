/*jslint nomen: true, browser:true*/
/*global define */

define(['spin', 'jquery'], function (Spinner, $) {
    "use strict";

    function WaitView(parentView) {
        this._parentView = parentView;
        this._spinView = $('<div class="full-parent"></div>');
        this._spinner = new Spinner(this._opts);
    }

    WaitView.prototype = {

        _opts: {
            lines: 11, // The number of lines to draw
            length: 20, // The length of each line
            width: 10, // The line thickness
            radius: 30, // The radius of the inner circle
            corners: 1, // Corner roundness (0..1)
            rotate: 0, // The rotation offset
            direction: 1, // 1: clockwise, -1: counterclockwise
            color: '#5C5C5C', // #rgb or #rrggbb or array of colors
            speed: 1, // Rounds per second
            trail: 60, // Afterglow percentage
            shadow: false, // Whether to render a shadow
            hwaccel: false, // Whether to use hardware acceleration
            className: 'spinner', // The CSS class to assign to the spinner
            zIndex: 2e9, // The z-index (defaults to 2000000000)
            top: 'auto', // Top position relative to parent in px
            left: 'auto' // Left position relative to parent in px
        },

        startWaiting: function () {
            this._spinView.appendTo(this._parentView);
            this._spinner.spin(this._spinView[0]);
        },
        stopWaiting: function () {
            this._spinner.stop();
            this._spinView.remove();
        }
    };

    return WaitView;

});