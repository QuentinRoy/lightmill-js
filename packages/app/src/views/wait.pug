.lightmill-base-app.lightmill-base-app-wait
  .lightmill-puck-wrapper
    .lightmill-puck.lightmill-puck-1
    .lightmill-puck.lightmill-puck-2
