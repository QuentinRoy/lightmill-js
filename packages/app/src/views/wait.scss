@import './common';

$puck-wrapper-size: 60px;
$puck-radius: 5px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .#{$class-prefix}puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    animation: $animation-name $rotation-duration infinite linear;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .#{$class-prefix}puck {
      position: absolute;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }

    .#{$class-prefix}puck-1 {
      top: 0;
      left: 0;
    }

    .#{$class-prefix}puck-2 {
      right: 0;
      bottom: 0;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
