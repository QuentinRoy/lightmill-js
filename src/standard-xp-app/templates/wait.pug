.xp-app-base.xp-app-base-wait
  .puck-wrapper
    .puck
    
