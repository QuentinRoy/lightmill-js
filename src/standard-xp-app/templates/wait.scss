@import './common';

$puck-wrapper-size: 400px;
$puck-radius: 25px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .puck {
      position: absolute;
      top: 0;
      left: 0;
      animation: $animation-name $rotation-duration infinite linear;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
