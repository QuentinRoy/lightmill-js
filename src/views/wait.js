import Spinner from 'spin';

const spinnerArgs = {
  lines: 11, // The number of lines to draw
  length: 20, // The length of each line
  width: 10, // The line thickness
  radius: 30, // The radius of the inner circle
  corners: 1, // Corner roundness (0..1)
  rotate: 0, // The rotation offset
  direction: 1, // 1: clockwise, -1: counterclockwise
  color: 'white', // #rgb or #rrggbb or array of colors
  speed: 1, // Rounds per second
  trail: 60, // Afterglow percentage
  shadow: false, // Whether to render a shadow
  hwaccel: false, // Whether to use hardware acceleration
  className: 'spinner', // The CSS class to assign to the spinner
  zIndex: 2e9, // The z-index (defaults to 2000000000)
  top: 'auto', // Top position relative to parent in px
  left: 'auto' // Left position relative to parent in px
};

export default class WaitView {
  constructor(parentNode) {
    this._parentNode = parentNode;
    this._spinDiv = document.createElement('div');
    Object.assign(this._spinDiv.style, {
      position: 'absolute',
      width: '100%',
      height: '100%',
      margin: 0,
      padding: 0,
      top: 0,
      left: 0
    });
    this._spinDiv.style.backgroundColor = 'black';
    this._spinner = new Spinner(spinnerArgs);
    this._isShown = false;
  }
  show() {
    if (!this.isShown) {
      this._parentNode.appendChild(this._spinDiv);
      this._spinner.spin(this._spinDiv);
      this._isShown = true;
    }
  }
  hide() {
    if (this.isShown) {
      this._spinner.stop();
      if (this._parentNode.contains(this._spinDiv)) {
        this._parentNode.removeChild(this._spinDiv);
      }
      this._isShown = false;
    }
  }
  get isShown() {
    return this._isShown;
  }
}
