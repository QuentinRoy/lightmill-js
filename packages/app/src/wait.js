/* eslint-disable no-param-reassign */

import waitTemplate from './views/wait.pug';
import './views/wait.scss';

export default (node, message) => {
  node.innerHTML = waitTemplate({ message });

  return {
    remove() {
      node.innerHTML = '';
    }
  };
};
