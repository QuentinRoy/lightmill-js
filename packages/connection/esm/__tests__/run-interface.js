import wait from 'wait-then';
import deferred from 'promise.defer';
import RunInterface, {
  connectToRun,
  selectRun,
  checkExperimentAndImportIfNeeded,
  consolidateRun
} from '../run-interface';

const spyDelay = (result, delay = 0) =>
  jest.fn(() => wait(delay).then(() => result));

describe('selectRun', () => {
  it('asks for an available run if no run is provided', async () => {
    const server = { availableRun: spyDelay({ experimentId: 'xpid' }) };
    expect(await selectRun(server, 'xpid')).toEqual({ experimentId: 'xpid' });
    expect(server.availableRun).toMatchSnapshot();
  });

  it('asks for the specified run if it is provided', async () => {
    const server = { run: spyDelay({ experimentId: 'xpid', id: 'runid' }) };
    expect(await selectRun(server, 'xpid', 'runid')).toEqual({
      experimentId: 'xpid',
      id: 'runid'
    });
    expect(server.run).toMatchSnapshot();
  });

  it('throws if the experiment ids do not match', async () => {
    const server = {
      availableRun: spyDelay({ id: 'runid', experimentId: 'xpid' }),
      run: spyDelay({ id: 'runid', experimentId: 'xpid' })
    };
    await expect(selectRun(server, 'other-xp-id')).rejects.toThrow(
      'Received experiment id is inconsistent with the one that has been requested.'
    );
  });

  it('throws if the run ids do not match', async () => {
    const server = {
      availableRun: spyDelay({ id: 'runid', experimentId: 'xpid' }),
      run: spyDelay({ id: 'runid', experimentId: 'xpid' })
    };
    await expect(selectRun(server, 'xpid', 'other-run-id')).rejects.toThrow(
      'Received run id is inconsistent with the one that has been requested.'
    );
  });
});

describe('connectToRun', () => {
  it("properly locks, ask for run's plan and for run's current trial when no run is specified", async () => {
    const server = {
      lock: spyDelay({ token: 'lock' }).mockName('server#lock'),
      plan: spyDelay(['block1', 'block2']).mockName('server#plan'),
      currentTrial: spyDelay({ number: 5, blockNumber: 10 }).mockName(
        'server#currentTrial'
      )
    };
    const result = await connectToRun(server, {
      id: 'runid',
      experimentId: 'xpid'
    });
    expect(server.lock).toMatchSnapshot();
    expect(server.plan).toMatchSnapshot();
    expect(server.currentTrial).toMatchSnapshot();
    expect(result).toEqual({
      id: 'runid',
      experimentId: 'xpid',
      blocks: ['block1', 'block2'],
      currentTrial: { number: 5, blockNumber: 10 },
      lock: { token: 'lock' }
    });
  });
});

describe('checkExperimentAndImportIfNeeded', () => {
  it('checks for an experiment on the server and does nothing if it is loaded', async () => {
    const loadedDef = deferred();
    const server = {
      isExperimentLoadedOnServer: jest.fn(() => loadedDef.promise)
    };
    const pr = checkExperimentAndImportIfNeeded(server, 'xp');
    expect(server.isExperimentLoadedOnServer.mock.calls).toEqual([['xp']]);
    loadedDef.resolve(true);
    await pr;
  });

  it('checks for an experiment on the server and load it if it is not loaded', async () => {
    const loadedDef = deferred();
    const importedDef = deferred();
    const server = {
      isExperimentLoadedOnServer: jest
        .fn(() => loadedDef.promise)
        .mockName('server#isExperimentLoadedOnServer'),
      importExperimentOnServer: jest
        .fn(() => importedDef.promise)
        .mockName('server#importExperimentOnServer')
    };
    const pr = checkExperimentAndImportIfNeeded(
      server,
      'xp',
      'http://host.path:1234/design.xml'
    );
    expect(server.isExperimentLoadedOnServer).toMatchSnapshot();
    expect(server.importExperimentOnServer).toMatchSnapshot();
    loadedDef.resolve(false);
    await wait();
    expect(server.importExperimentOnServer).toMatchSnapshot();
    importedDef.resolve();
    await pr;
  });

  it('fails without trying to import if it is not loaded and its design URI is not provided', async () => {
    const server = {
      isExperimentLoadedOnServer: spyDelay(false).mockName(
        'server#isExperimentLoadedOnServer'
      ),
      importExperimentOnServer: jest
        .fn()
        .mockName('server#importExperimentOnServer')
    };
    const pr = checkExperimentAndImportIfNeeded(server, 'xp');
    await expect(pr).rejects.toThrow();
    expect(server.isExperimentLoadedOnServer).toMatchSnapshot();
    expect(server.importExperimentOnServer).toMatchSnapshot();
  });
});

const makeRunPlan = () => ({
  id: 'runid',
  experimentId: 'xpid',
  blocks: [
    {
      values: { bval1: 'b11', bval2: 'b12' },
      number: 0,
      trials: [
        { values: { tval1: 't111', tval2: 't112' }, number: 0 },
        { values: { tval1: 't121', tval2: 't122' }, number: 1 }
      ]
    },
    {
      values: { bval1: 'b21', bval2: 'b22' },
      number: 1,
      trials: [
        { values: { tval1: 't211', tval2: 't212' }, number: 0 },
        { values: { tval1: 't221', tval2: 't222' }, number: 1 }
      ]
    }
  ]
});

describe('consolidateRun', () => {
  it('properly inserts backrefs.', () => {
    expect.assertions(6);
    const run = consolidateRun(makeRunPlan());
    run.blocks.forEach(block => {
      expect(block.run).toBe(run);
      block.trials.forEach(trial => {
        expect(trial.block).toBe(block);
      });
    });
  });
});

const makeRCon = ({
  run = consolidateRun(makeRunPlan()),
  token = '',
  trial = 0,
  block = 0,
  queue = {
    push: spyDelay(),
    flush: spyDelay()
  },
  post = spyDelay()
} = {}) => ({
  run,
  queue,
  post,
  rcon: new RunInterface(run, token, block, trial, queue, post)
});

describe('RunInterface', () => {
  it('finds the first trial', async () => {
    const { rcon, run } = makeRCon({ trial: 0, block: 1 });
    expect(run.blocks[1].trials[0]).toBe(await rcon.getCurrentTrial());
    expect(await rcon.getCurrentBlock()).toBe(run.blocks[1]);
  });

  it('finds the next trial when on the same block', async () => {
    const { rcon } = makeRCon({ trial: 0, block: 1 });
    const next = await rcon.getNextTrial();
    expect(next.number).toBe(1);
    expect(next.block.number).toBe(1);
  });

  it('finds the next trial when on the next block', async () => {
    const { rcon } = makeRCon({ trial: 1, block: 0 });
    const next = await rcon.getNextTrial();
    expect(next.number).toBe(0);
    expect(next.block.number).toBe(1);
  });

  it('#getNextTrial resolves with undefined when on the last trial', async () => {
    const { rcon } = makeRCon({ trial: 1, block: 1 });
    expect(await rcon.getNextTrial()).toBe(undefined);
  });

  it('#postResults properly posts the measures', async () => {
    const { rcon, post } = makeRCon({ token: 'token' });
    await rcon.postResults({ val: 'val' });
    expect(post).toMatchSnapshot();
  });

  it('#postResults throws if called twice on the same trial', async () => {
    const { rcon } = makeRCon();
    rcon.postResults({ val: 'val' });
    await expect(rcon.postResults({ val: 'other val' })).rejects.toThrow();
  });

  it('#postResults throws if post failed', async () => {
    const { rcon } = makeRCon({
      post: jest.fn(() =>
        wait().then(() => {
          throw new Error('test err');
        })
      )
    });
    await expect(rcon.postResults({ m: 'measure' })).rejects.toThrow(
      'test err'
    );
  });

  it('#endTrial throws if no results have been posted', async () => {
    const { rcon } = makeRCon({ trial: 0, block: 1 });
    await expect(rcon.endTrial()).rejects.toThrow();
  });

  it('#endTrial properly switches to the next trial', async () => {
    const { rcon } = makeRCon({ trial: 0, block: 1 });
    const currentTrial = await rcon.getCurrentTrial();
    const nextTrial = await rcon.getNextTrial();
    expect(currentTrial.number).toBe(0);
    expect(currentTrial.block.number).toBe(1);
    expect(nextTrial.number).toBe(1);
    expect(nextTrial.block.number).toBe(1);
    await rcon.postResults({});
    expect(await rcon.endTrial()).toBe(nextTrial);
    expect(await rcon.getCurrentTrial()).toBe(nextTrial);
  });

  it('#postResults posts the results sequentially', async () => {
    const post = jest.fn().mockName('post');
    const defs = Array.from({ length: 3 }).map(() => deferred());
    defs.forEach(def => post.mockReturnValueOnce(def.promise));
    const { rcon } = makeRCon({ trial: 0, block: 0, post, token: 'token' });
    rcon.postResults({ val: 'val1' });
    await rcon.endTrial();
    rcon.postResults({ val: 'val2' });
    await rcon.endTrial();
    rcon.postResults({ val: 'val3' });
    await rcon.endTrial();
    // Only one call for now as no post as returned yet.
    expect(post).toMatchSnapshot();
    defs[0].resolve();
    await wait();
    // Second call call now.
    expect(post).toMatchSnapshot();
    defs[1].resolve();
    await wait();
    // Third call call now.
    expect(post).toMatchSnapshot();
  });

  it('#endTrial resolves even if postResults is not done', async () => {
    const def = deferred();
    const post = jest.fn(() => def.promise);
    const { rcon } = makeRCon({ post });
    let postDone = false;
    rcon.postResults({ val: 'val1' }).then(() => {
      postDone = true;
    });
    await rcon.endTrial();
    expect(postDone).toBe(false);
  });

  it('#flush is delegated on the queue', async () => {
    const def = deferred();
    const queue = {
      push: spyDelay(),
      flush: jest.fn(() => def.promise)
    };
    const { rcon } = makeRCon({ queue });
    let resolved = false;
    rcon.flush(42).then(() => {
      resolved = true;
    });
    expect(queue.flush).toMatchSnapshot();
    await wait();
    expect(resolved).toBe(false);
    def.resolve();
    await wait();
    expect(resolved).toBe(true);
  });
});
