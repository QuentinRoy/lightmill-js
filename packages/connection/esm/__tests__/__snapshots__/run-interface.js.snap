// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunInterface #flush is delegated on the queue 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      42,
    ],
  ],
}
`;

exports[`RunInterface #postResults posts the results sequentially 1`] = `
[MockFunction post] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
      0,
      0,
      Object {
        "measures": Object {
          "val": "val1",
        },
        "token": "token",
      },
    ],
  ],
}
`;

exports[`RunInterface #postResults posts the results sequentially 2`] = `
[MockFunction post] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
      0,
      0,
      Object {
        "measures": Object {
          "val": "val1",
        },
        "token": "token",
      },
    ],
    Array [
      "xpid",
      "runid",
      0,
      1,
      Object {
        "measures": Object {
          "val": "val2",
        },
        "token": "token",
      },
    ],
  ],
}
`;

exports[`RunInterface #postResults posts the results sequentially 3`] = `
[MockFunction post] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
      0,
      0,
      Object {
        "measures": Object {
          "val": "val1",
        },
        "token": "token",
      },
    ],
    Array [
      "xpid",
      "runid",
      0,
      1,
      Object {
        "measures": Object {
          "val": "val2",
        },
        "token": "token",
      },
    ],
    Array [
      "xpid",
      "runid",
      1,
      0,
      Object {
        "measures": Object {
          "val": "val3",
        },
        "token": "token",
      },
    ],
  ],
}
`;

exports[`RunInterface #postResults properly posts the measures 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
      0,
      0,
      Object {
        "measures": Object {
          "val": "val",
        },
        "token": "token",
      },
    ],
  ],
}
`;

exports[`checkExperimentAndImportIfNeeded checks for an experiment on the server and load it if it is not loaded 1`] = `
[MockFunction server#isExperimentLoadedOnServer] {
  "calls": Array [
    Array [
      "xp",
    ],
  ],
}
`;

exports[`checkExperimentAndImportIfNeeded checks for an experiment on the server and load it if it is not loaded 2`] = `[MockFunction server#importExperimentOnServer]`;

exports[`checkExperimentAndImportIfNeeded checks for an experiment on the server and load it if it is not loaded 3`] = `
[MockFunction server#importExperimentOnServer] {
  "calls": Array [
    Array [
      "http://host.path:1234/design.xml",
    ],
  ],
}
`;

exports[`checkExperimentAndImportIfNeeded fails without trying to import if it is not loaded and its design URI is not provided 1`] = `
[MockFunction server#isExperimentLoadedOnServer] {
  "calls": Array [
    Array [
      "xp",
    ],
  ],
}
`;

exports[`checkExperimentAndImportIfNeeded fails without trying to import if it is not loaded and its design URI is not provided 2`] = `[MockFunction server#importExperimentOnServer]`;

exports[`connectToRun properly locks, ask for run's plan and for run's current trial when no run is specified 1`] = `
[MockFunction server#lock] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
    ],
  ],
}
`;

exports[`connectToRun properly locks, ask for run's plan and for run's current trial when no run is specified 2`] = `
[MockFunction server#plan] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
    ],
  ],
}
`;

exports[`connectToRun properly locks, ask for run's plan and for run's current trial when no run is specified 3`] = `
[MockFunction server#currentTrial] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
    ],
  ],
}
`;

exports[`selectRun asks for an available run if no run is provided 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "xpid",
    ],
  ],
}
`;

exports[`selectRun asks for the specified run if it is provided 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "xpid",
      "runid",
    ],
  ],
}
`;
